lib/charms/alertmanager_k8s/v0/alertmanager_remote_configuration.py

```diff
@@ -0,0 +1,459 @@
+
+    ```
+    requires:
+        remote-configuration:  # Relation name
+            interface: alertmanager_remote_configuration  # Relation interface
+            limit: 1
+    ```
+
+    The `RemoteConfigurationRequirer` provides a public `config` method for exposing the data
+    from the relation data bag. Typical usage of these methods in the provider charm would look
+    something like:
+
+    ```
+    def get_config(self, *args):
+        ...
+        configuration, templates = self.remote_configuration.config()
+        ...
+        self.container.push("/alertmanager/config/file.yml", configuration)
+        self.container.push("/alertmanager/templates/file.tmpl", templates)
+        ...
+    ```
+
+    Separation of the main configuration and the templates is dictated by the assumption that
+    the default provider of the `alertmanager_remote_configuration` relation will be
+    `alertmanager-k8s` charm, which requires such separation.
+    """
+
+    on = AlertmanagerRemoteConfigurationRequirerEvents()
+
+    def __init__(
+        self,
+        charm: CharmBase,
+        relation_name: str = DEFAULT_RELATION_NAME,
+    ):
+        """API that manages a required `remote-configuration` relation.
+
+        Args:
+            charm: The charm object that instantiated this class.
+            relation_name: Name of the relation with the `alertmanager_remote_configuration`
+                interface as defined in metadata.yaml. Defaults to `remote-configuration`.
+        """
+        super().__init__(charm, relation_name)
+        self._charm = charm
+        self._relation_name = relation_name
+
+        on_relation = self._charm.on[self._relation_name]
+
+        self.framework.observe(on_relation.relation_created, self._on_relation_created)
+        self.framework.observe(on_relation.relation_changed, self._on_relation_changed)
+        self.framework.observe(on_relation.relation_broken, self._on_relation_broken)
+
+    def _on_relation_created(self, _) -> None:
+        """Event handler for remote configuration relation created event.
+
+        Informs about the fact that the configuration from remote provider will be used.
+        """
+        logger.debug("Using remote configuration from the remote_configuration relation.")
+
+    def _on_relation_changed(self, _) -> None:
+        """Event handler for remote configuration relation changed event.
+
+        Emits custom `remote_configuration_changed` event every time remote configuration
+        changes.
+        """
+        self.on.remote_configuration_changed.emit()
+
+    def _on_relation_broken(self, _) -> None:
+        """Event handler for remote configuration relation broken event.
+
+        Informs about the fact that the configuration from remote provider will no longer be used.
+        """
+        logger.debug("Remote configuration no longer available.")
+
+    def config(self) -> Tuple[Optional[dict], Optional[list]]:
```

Comment:
Shouldn't this be:
```suggestion
    def config(self) -> Optional[Tuple[dict, list]]:
```
I.e. `({"some": "thing"}, [])` rather than `({"some": "thing"}, None)`?

(This is not a blocking comment.)

Reply:
The signature matches current implementation - both `_alertmanager_config` and `_alertmanager_templates` return `None` if there's no data in the relation data bag. If you see a benefit of returning an empty dict/list rather than `None`, I'm open to changing the implementation.

Reply:
With the current signature, config may be expected to return None for config and non-None for templates.
Seems to me that the `Optional[Tuple[dict, list]]` signatures better conveys intent.
But again, this is minor and can be addressed later. Let's get this merged already :)

Reply:
Lol :D No can do, unless @simskij removes change request ;)